ugo.go

```diff
@@ -0,0 +1,223 @@
+	}
+}
+
+func get_possible_package_names() (pkgs_to_path map[string]string, err error) {
+	current_dir, err := os.Getwd()
+	pkgs_to_path = map[string]string{}
+
+	if err != nil {
+		return pkgs_to_path, err
+	}
+
+	for ; current_dir != ""; current_dir = get_parent_path(current_dir) {
+		repo, err := git.PlainOpen(current_dir)
+
+		if err != nil {
+			continue
+		}
+
+		conf, err := repo.Config()
+
+		if err != nil {
+			continue
+		}
+
+		for _, remote := range conf.Remotes {
+			url, err := git_url_to_package_name(remote.URL)
+
+			if err != nil {
+				continue
+			}
+
+			pkgs_to_path[url] = current_dir
+		}
+	}
+
+	return pkgs_to_path, nil
+}
+
+func create_gopackage_file() (string, error) {
+	pkgs_to_path, err := get_possible_package_names()
+
+	if err != nil {
+		return "", err
+	}
+
+	for pkg_name, pkg_path := range pkgs_to_path {
+		if prompt.Confirm("Do you want to create a %s/%s file with the following content: %s", pkg_path, GOPKG_FILENAME, pkg_name) {
+			err := ioutil.WriteFile(path.Join(pkg_path, GOPKG_FILENAME), []byte(pkg_name), 0644)
+
+			return pkg_path, err
+		}
+	}
+
+	if len(pkgs_to_path) > 0 {
+		return "", fmt.Errorf("No package selected to create %s file, please create the file manually.", GOPKG_FILENAME)
+	} else {
+		return "", fmt.Errorf("No package auto detected to create %s file, please create the file manually.", GOPKG_FILENAME)
+	}
+}
+
+func create_workspace(gopkg_path string) error {
+	package_name, err := get_gopackage(gopkg_path)
+
+	if err != nil {
+		return err
+	}
+
+	base_path, dir := path.Split(package_name)
+
+	workspace_path := path.Join(gopkg_path, ".workspace", "src", base_path)
+
+	os.MkdirAll(workspace_path, os.ModePerm)
+
+	symlink := path.Join(workspace_path, dir)
+
+	if path_exists(symlink) {
+		os.Remove(symlink)
+	}
+
+	os.Symlink(gopkg_path, symlink)
+
+	return nil
+}
+
+func get_environment(replace bool, gopkg_path string) []string {
+	env := make(map[string]string)
+	for _, e := range os.Environ() {
+		if i := strings.Index(e, "="); i >= 0 {
+			env[e[:i]] = e[i+1:]
+		}
+	}
+
+	gopath := path.Join(gopkg_path, ".workspace")
+
+	if replace {
+		env["GOPATH"] = gopath
+	} else {
+		env["GOPATH"] += ":" + gopath
+	}
+
+	result := []string{}
+
+	for key, value := range env {
+		result = append(result, key+"="+value)
+	}
+
+	return result
+}
+
+var (
+	replace_gopath = kingpin.Flag("replace", "Replace existing GOPATH instead of adding.").Default("false").Short('r').Bool()
+	command        = kingpin.Arg("command", "The command that should be executed by ugo.").Required().String()
+	command_args   = kingpin.Arg("arguments", "The arguments for the command.").Strings()
+)
+
+func main() {
+	kingpin.Parse()
+
+	gopkg_path, err := find_gopackage_path()
+
+	if err != nil {
+		gopkg_path, err = create_gopackage_file()
+
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	}
+
+	err = create_workspace(gopkg_path)
+
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	cmdstr := *command
```

Comment:
This could also be written as
```
cmdlist := append([]string{*command}, (*command_args)...)
cmdstr := strings.Join(cmdlist, " ")
```
It's shorter, but maybe not more readable.